randovania/games/dread/layout/dread_damage_state.py

```diff
@@ -0,0 +1,111 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Self, override
+
+from randovania.game_description.requirements.resource_requirement import ResourceRequirement
+from randovania.resolver.energy_tank_damage_state import EnergyTankDamageState
+
+if TYPE_CHECKING:
+    from collections.abc import Generator
+
+    from randovania.game_description.resources.item_resource_info import ItemResourceInfo
+    from randovania.game_description.resources.resource_collection import ResourceCollection
+
+
+class DreadDamageState(EnergyTankDamageState):
+    _use_immediate_energy_parts: bool
+    _energy_part_item: ItemResourceInfo
+
+    def __init__(
+        self,
+        starting_energy: int,
+        energy_per_tank: int,
+        energy_tank: ItemResourceInfo,
+        use_immediate_energy_parts: bool,
+        energy_part_item: ItemResourceInfo,
+    ):
+        super().__init__(starting_energy, energy_per_tank, energy_tank)
+        self._use_immediate_energy_parts = use_immediate_energy_parts
+        self._energy_part_item = energy_part_item
+
+    def __copy__(self) -> Self:
+        return self._duplicate()
+
+    @override
+    def resources_for_energy(self) -> Generator[ItemResourceInfo]:
+        yield self._energy_tank
+        yield self._energy_part_item
+
+    def _maximum_energy(self, resources: ResourceCollection) -> int:
+        num_tanks = resources[self._energy_tank]
+        num_parts = resources[self._energy_part_item]
+        num_whole_tanks = num_tanks + num_parts // 4
+        num_partial_parts = num_parts % 4
+        energy_from_partial = (self._energy_per_tank * num_partial_parts) // 4
+
+        return self._starting_energy + (self._energy_per_tank * num_whole_tanks) + energy_from_partial
+
+    def _duplicate(self) -> Self:
+        result = DreadDamageState(
```

Comment:
this should be `self.__class__(` rather than `DreadDamageState(` to achieve type safety (or you can use a type ignore directive)